### PR TITLE
symbolic.py: faster Node.sum, faster SumNode.div

### PR DESCRIPTION
**Pull request by roelofvandijk**
Minor refactor of symbolic.py

1) Simplify node partitioning in `Node.sum`
2) Replace `isinstance` with faster `__class__` check where possible.
3) Replace use of partition helper function with faster oneliner

I believe the timing gains of the first two are of the order of the measurement noise, the last change is however quite a bit faster.

Measured using `kernprof -lv pytest -s test/test_speed_v_torch.py`

`Node.sum`
Before
```
Total time: 0.715114 s
File: tinygrad/tinygrad/shape/symbolic.py
Function: sum at line 60

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    60                                             @staticmethod
    61                                             @profile
    62                                             def sum(nodes:List[Node]) -> Node:
    63     25740       8056.8      0.3      1.1      new_nodes: List[Node] = []
    64     25740       6732.8      0.3      0.9      sum_nodes: List[SumNode] = []
    65     25740       6307.9      0.2      0.9      num_nodes: List[NumNode] = []
    66     25740       5825.5      0.2      0.8      mul_nodes: List[MulNode] = []
    67     91400      28279.7      0.3      4.0      for node in nodes:
    68     47923      17353.2      0.4      2.4        if isinstance(node, NumNode):
    69     43477      15765.7      0.4      2.2          num_nodes.append(node)
    70     37626      11982.1      0.3      1.7        elif isinstance(node, MulNode):
    71     37626      13254.2      0.4      1.9          mul_nodes.append(node)
    72      7837       2498.4      0.3      0.3        elif isinstance(node, SumNode): # expand any sums inside one sum
    73      2460        978.9      0.4      0.1          sum_nodes.append(node)
    74                                                 else:
    75      7837       2957.3      0.4      0.4          new_nodes.append(node)
    76                                           
    77                                               # expand any sums inside one sum
    78     24400       7164.3      0.3      1.0      if sum_nodes:
    79      1340        549.9      0.4      0.1        new_nodes.extend(num_nodes)
    80      1340        426.4      0.3      0.1        new_nodes.extend(mul_nodes)
    81      2460       1184.5      0.5      0.2        for x in sum_nodes: new_nodes += x.nodes
    82      1340       1688.4      1.3      0.2        return Variable.sum(new_nodes)
    83                                           
    84                                               # combine any numbers inside a sum
    85     22617       5451.7      0.2      0.8      if num_nodes:
    86     22617      41027.0      1.8      5.7        new_nodes.append(NumNode(sum([x.b for x in num_nodes])))
    87                                           
    88                                               # combine any MulNodes that factorize (big hack sticking the MulNode(x, 1) on things)
    89     24400      61250.7      2.5      8.6      mul_nodes += [MulNode(x, 1) for x in new_nodes]
    90     24400      14743.3      0.6      2.1      mul_groups: Dict[str, Tuple[Node, List[MulNode]]] = defaultdict(lambda: (Node(), []))
    91     67806      17835.3      0.3      2.5      for node in mul_nodes: #NOTE can we somehow avoid rendering here?
    92     67806      93136.9      1.4     13.0        key = node.a.render()
    93     67806      67779.6      1.0      9.5        mul_groups[key] = (node.a, mul_groups[key][1] + [node])
    94     24400     164845.8      6.8     23.1      mul_nodes = [k * sum(x.b for x in g) for k, g in mul_groups.values()]
    95     24400      32100.9      1.3      4.5      new_nodes = [x if not isinstance(x, MulNode) or x.b != 1 else x.a for x in mul_nodes]
    96                                           
    97                                               # filter 0s
    98     24400      29772.2      1.2      4.2      new_nodes = [x for x in new_nodes if x.min != 0 or x.max != 0]
    99     24400      56164.9      2.3      7.9      return create_rednode(SumNode, new_nodes) if len(new_nodes) > 1 else (new_nodes[0] if len(new_nodes) == 1 else NumNode(0))
```

After:
```
Total time: 0.713905 s
File: tinygrad/tinygrad/shape/symbolic.py
Function: sum at line 60

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    60                                             @staticmethod
    61                                             @profile
    62                                             def sum(nodes:List[Node]) -> Node:
    63     25740      14202.0      0.6      2.0      new_nodes, num_nodes, mul_nodes, sum_nodes = [],[],[],[]
    64     25740      13887.1      0.5      1.9      lists = {NumNode : num_nodes, MulNode : mul_nodes, SumNode : sum_nodes}
    65     90790      25728.6      0.3      3.6      for node in nodes:
    66     90790      48282.1      0.5      6.8        lists.get(node.__class__, new_nodes).append(node)
    67                                           
    68                                               # expand any sums inside one sum
    69     24400       7378.9      0.3      1.0      if sum_nodes:
    70      1340        541.1      0.4      0.1        new_nodes.extend(num_nodes)
    71      1340        437.6      0.3      0.1        new_nodes.extend(mul_nodes)
    72      2460       1216.6      0.5      0.2        for x in sum_nodes: new_nodes += x.nodes
    73      1340       1965.5      1.5      0.3        return Variable.sum(new_nodes)
    74                                           
    75                                               # combine any numbers inside a sum
    76     22617       6167.3      0.3      0.9      if num_nodes:
    77     22617      42988.4      1.9      6.0        new_nodes.append(NumNode(sum([x.b for x in num_nodes])))
    78                                           
    79                                               # combine any MulNodes that factorize (big hack sticking the MulNode(x, 1) on things)
    80     24400      62585.3      2.6      8.8      mul_nodes += [MulNode(x, 1) for x in new_nodes]
    81     24400      15714.6      0.6      2.2      mul_groups: Dict[str, Tuple[Node, List[MulNode]]] = defaultdict(lambda: (Node(), []))
    82     67806      19338.9      0.3      2.7      for node in mul_nodes: #NOTE can we somehow avoid rendering here?
    83     67806      97178.8      1.4     13.6        key = node.a.render()
    84     67806      71796.2      1.1     10.1        mul_groups[key] = (node.a, mul_groups[key][1] + [node])
    85     24400     165823.0      6.8     23.2      mul_nodes = [k * sum(x.b for x in g) for k, g in mul_groups.values()]
    86     24400      32989.6      1.4      4.6      new_nodes = [x if not x.__class__ is MulNode or x.b != 1 else x.a for x in mul_nodes]
    87                                           
    88                                               # filter 0s
    89     24400      29561.6      1.2      4.1      new_nodes = [x for x in new_nodes if x.min != 0 or x.max != 0]
    90     24400      56121.9      2.3      7.9      return create_rednode(SumNode, new_nodes) if len(new_nodes) > 1 else (new_nodes[0] if len(new_nodes) == 1 else NumNode(0))
```

`__div__`
Before:
```
Total time: 0.76355 s
File: tinygrad/tinygrad/shape/symbolic.py
Function: __floordiv__ at line 169

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   169                                             @profile
   170                                             def __floordiv__(self, b: int, factoring_allowed=True):
   171     11932       3460.8      0.3      0.5      if not factoring_allowed: return Node.__floordiv__(self, b, factoring_allowed)
   172     11932      70917.6      5.9      9.3      factors, tmp_nofactor = partition(self.nodes, lambda x: (isinstance(x, (MulNode, NumNode))) and x.b%b == 0)
   173     11932       3137.5      0.3      0.4      nofactor = []
   174                                               # ugh, i doubt this is universally right
   175     30508       8548.6      0.3      1.1      for x in tmp_nofactor:
   176     19977       7129.3      0.4      0.9        if isinstance(x, NumNode):
   177      9921       3726.5      0.4      0.5          if (x.b%b) != x.b:
   178       610        899.4      1.5      0.1            factors.append(Variable.num(x.b - (x.b%b)))  # python does floor division
   179     10531      13414.6      1.3      1.8          nofactor.append(Variable.num(x.b%b))
   180                                                 else:
   181     19977       7202.3      0.4      0.9          nofactor.append(x)
   182     11932      22465.9      1.9      2.9      gcd = [math.gcd(x.b, b) if isinstance(x, (MulNode, NumNode)) else None for x in nofactor]
   183      7035       2796.9      0.4      0.4      if len(factors) > 0:
   184                                                 # these don't have to be the same, just having a common factor
   185      5758      13281.4      2.3      1.7        if len(gcd) > 0 and all_same(gcd) and gcd[0] is not None and gcd[0] > 1:
   186      1185      37760.1     31.9      4.9          nofactor_term = Variable.sum([(x.a * (x.b//gcd[0])) if isinstance(x, MulNode) else Variable.num(x.b//gcd[0]) for x in nofactor])//(b//gcd[0])
   187                                                 else:
   188      4029     203690.7     50.6     26.7          nofactor_term = Variable.sum(nofactor)//b
   189      7035     348552.9     49.5     45.6        return Variable.sum([(x.a * (x.b//b)) if isinstance(x, MulNode) else Variable.num(x.b//b) for x in factors] + [nofactor_term])
   190                                               else:
   191      4897       6141.9      1.3      0.8        muls = [x.b for x in nofactor if isinstance(x, MulNode)]
   192      4709       1376.6      0.3      0.2        for m in muls:
   193      4309       1747.2      0.4      0.2          if m > 1 and b%m == 0:
   194      4309       5289.6      1.2      0.7            return (self//m)//(b//m)
   195       588       2010.3      3.4      0.3        return Node.__floordiv__(self, b, factoring_allowed)
```

After:
```
Total time: 0.650139 s
File: tinygrad/tinygrad/shape/symbolic.py
Function: __floordiv__ at line 160

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   160                                             @profile
   161                                             def __floordiv__(self, b: int, factoring_allowed=True):
   162     11932       3383.3      0.3      0.5      if not factoring_allowed: return Node.__floordiv__(self, b, factoring_allowed)
   163     11932       3832.1      0.3      0.6      factors, tmp_nofactor, nofactor = [], [], []
   164     44852      31764.2      0.7      4.9      for x in self.nodes: factors.append(x) if x.__class__ in (MulNode, NumNode) and x.b%b == 0 else nofactor.append(x)**
   165                                               # ugh, i doubt this is universally right
   166     11932       3673.2      0.3      0.6      for x in tmp_nofactor:
   167                                                 if x.__class__ is NumNode:
   168                                                   if (x.b%b) != x.b:
   169                                                     factors.append(Variable.num(x.b - (x.b%b)))  # python does floor division
   170                                                   nofactor.append(Variable.num(x.b%b))
   171                                                 else:
   172                                                   nofactor.append(x)
   173     11932      22023.4      1.8      3.4      gcd = [math.gcd(x.b, b) if  x.__class__ in (MulNode, NumNode) else None for x in nofactor]
   174      7035       2873.0      0.4      0.4      if len(factors) > 0:
   175                                                 # these don't have to be the same, just having a common factor
   176      5758      14024.7      2.4      2.2        if len(gcd) > 0 and all_same(gcd) and gcd[0] is not None and gcd[0] > 1:
   177      1185      36181.0     30.5      5.6          nofactor_term = Variable.sum([(x.a * (x.b//gcd[0])) if x.__class__ is MulNode else Variable.num(x.b//gcd[0]) for x in nofactor])//(b//gcd[0])
   178                                                 else:
   179      4029     191465.6     47.5     29.4          nofactor_term = Variable.sum(nofactor)//b
   180      7035     323973.8     46.1     49.8        return Variable.sum([(x.a * (x.b//b)) if x.__class__ is MulNode else Variable.num(x.b//b) for x in factors] + [nofactor_term])
   181                                               else:
   182      4897       6274.3      1.3      1.0        muls = [x.b for x in nofactor if x.__class__ is MulNode]
   183      4709       1470.5      0.3      0.2        for m in muls:
   184      4309       1886.4      0.4      0.3          if m > 1 and b%m == 0:
   185      4309       5048.2      1.2      0.8            return (self//m)//(b//m)
   186       588       2265.7      3.9      0.3        return Node.__floordiv__(self, b, factoring_allowed)
```

### CODE REVIEW

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/shape/symbolic.py                14              23
------------------------------------------------------------
lines added in the tinygrad folder: -9
```

File: tinygrad/shape/symbolic.py [hunk 1]
```diff
@@ -59,19 +59,10 @@ def num(num:int) -> Node: return NumNode(num)
 
   @staticmethod
   def sum(nodes:List[Node]) -> Node:
-    new_nodes: List[Node] = []
```
Comment (geohot) [inline tinygrad/shape/symbolic.py:62]: Why remove the type info here

**Comment by roelofvandijk** [comment]
New faster sum implementation. Change to draft MR.

- Loop over nodes once
- Add flatten property to SumNodes
- Only factoring when necessary
- Simplify factoring code

These functions get hit a lot and could be cached, but I think that the current hashing of nodes is slow and could be improved. Therefore most caching here is not actually speeding things up.

```
Total time: 0.482461 s
File: /home/rvd/src/roelofvandijk/tinygrad/tinygrad/shape/symbolic.py
Function: sum at line 70

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    70                                             @staticmethod
    71                                             @profile
    72                                             def sum(nodes:List[Node]) -> Node:
    73     23565       7477.1      0.3      1.5      if not nodes: return NumNode(0)
    74     21407      10390.8      0.5      2.2      if len(nodes) == 1: return nodes[0]
    75                                           
    76     21407       6138.1      0.3      1.3      new_nodes: List[Node] = []
    77     21407       4953.4      0.2      1.0      a_nodes: List[Node] = []
    78     21407       5422.6      0.3      1.1      num_node_sum = 0
    79                                           
    80     85482      25509.5      0.3      5.3      for node in nodes:
    81     83022      32155.7      0.4      6.7        if isinstance(node, SumNode):
    82      2460      10407.0      4.2      2.2          flat = node.flat_components_grouped_num  # last component is always numnode
    83      2460       1327.0      0.5      0.3          nodes += flat[:-1]
    84      2460        924.1      0.4      0.2          num_node_sum += flat[-1].b
    85     45122      14828.4      0.3      3.1        elif node.__class__ is NumNode: 
    86     37900      12717.9      0.3      2.6          num_node_sum += node.b
    87                                                 else: 
    88     45122      17322.7      0.4      3.6          new_nodes.append(node)
    89     45122      22108.1      0.5      4.6          if isinstance(node, MulNode): a_nodes.append(node.a)
    90                                           
    91     11260       3064.0      0.3      0.6      if not num_node_sum:
    92      9317       2342.5      0.3      0.5        if not new_nodes: return NumNode(0)
    93      6465       3155.0      0.5      0.7        if len(new_nodes) == 1: return new_nodes[0]
    94     13516       3428.7      0.3      0.7      if not new_nodes: return NumNode(num_node_sum)
    95                                               
    96     13516     231653.4     17.1     48.0      if len(set(a_nodes)) < len(nodes): new_nodes = Node.factor(tuple(new_nodes)) # only factor if duplicate a-nodes exist
    97                                           
    98     13516      11530.4      0.9      2.4      if num_node_sum: new_nodes.append(NumNode(num_node_sum))  
    99                                           
   100     13516      55605.0      4.1     11.5      return create_rednode(SumNode, new_nodes) if len(new_nodes) > 1 else new_nodes[0] if len(new_nodes) == 1 else NumNode(0)
```

**Comment by roelofvandijk** [comment]
```
Total time: 0.222576 s
File: /home/rvd/src/roelofvandijk/tinygrad/tinygrad/shape/symbolic.py
Function: __floordiv__ at line 166

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   166                                             @profile
   167                                             def __floordiv__(self, b: int, factoring_allowed=True):
   168     11600       3523.2      0.3      1.6      if b == 1: return self
   169     11600       2902.6      0.3      1.3      if not factoring_allowed: return Node.__floordiv__(self, b, factoring_allowed)
   170     11600       2981.1      0.3      1.3      factors: List[Node] = []
   171     11600       2622.6      0.2      1.2      nofactor_mul: List[Node] = []
   172     11600       2690.3      0.2      1.2      nofactor_nonmul: List[Node] = []
   173     42755      30226.6      0.7     13.6      for x in self.flat_components: 
   174     42381      13452.0      0.3      6.0        if x.__class__ is NumNode and x.b%b == 0: factors.append(x)
   175     31130      17135.6      0.6      7.7        elif x.__class__ is MulNode: factors.append(x) if x.b%b == 0 else  nofactor_mul.append(x) 
   176     11251       3952.1      0.4      1.8        else: nofactor_nonmul.append(x)
   177                                                 
   178      6675       1694.7      0.3      0.8      if factors:  # factor out largest possible gcd
   179      6675      51426.2      7.7     23.1        factor_term = [x.a * x.b//b if isinstance(x, MulNode) else NumNode(x.b//b) for x in factors]
   180      6600       1830.5      0.3      0.8        if nofactor_mul and not nofactor_nonmul:
   181        75        115.3      1.5      0.1          gcds = [gcd(x.b, b) for x in nofactor_mul]
   182        75        158.3      2.1      0.1          if (t := min(gcds)) > 1 and all([x.b%t == 0 for x in nofactor_mul]): 
   183        51       1130.4     22.2      0.5            nofactor_term = [Node.sum([x.a * x.b//t for x in nofactor_mul if isinstance(x, MulNode)])//(b//t)]  # mypy wants the isinstance
   184                                                   else:
   185                                                     nofactor_term = [Node.sum(nofactor_mul)//b] if nofactor_mul else []
   186                                                 else: 
   187      4070      39456.4      9.7     17.7          nofactor_term = [Node.sum(nofactor_mul+nofactor_nonmul)//b] if nofactor_mul + nofactor_nonmul else []
   188      6675      38250.7      5.7     17.2        return Node.sum(factor_term + nofactor_term)
   189      4737       1270.8      0.3      0.6      for m in nofactor_mul:
   190      4337       5855.6      1.4      2.6        if m.b > 1 and b%m.b == 0: return (self//m.b)//(b//m.b)
   191       588       1900.9      3.2      0.9      return Node.__floordiv__(self, b, factoring_allowed)
```

File: tinygrad/shape/symbolic.py [hunk 2]
```diff
@@ -57,49 +56,48 @@ def __mod__(self, b:int):
   @staticmethod
   def num(num:int) -> Node: return NumNode(num)
 
+
+  @staticmethod
+  @functools.lru_cache(maxsize=None)
+  def factorize(nodes: List[Node]):
```
Comment (geohot) [inline tinygrad/shape/symbolic.py:62]: Can you mark the return type here?

**Comment by geohot** [comment]
Looks good! Minor nits in comments.

Btw, there's a fuzzer test for this: `python3 test/external/fuzz_symbolic.py` ran it and it looked good.

**Comment by roelofvandijk** [comment]
Thanks! Should be good to go. There's an efficient specialization for the case n=2, but I can add that later (if 2 different types and no SumNode -> create SumNode).

New timing should be around 0.2 for the function. Caching the factorization did not help so I left it out. I think that the current hashing of the nodes is not efficient, which makes caching operations here not efficient.

**Comment by roelofvandijk** [comment]
It would be very nice if there was a dedicated CI/CD test that evaluates a representative performance, e.g. for LLAMA.

For now I have been using test_net_speed, test_speed_v_torch and the llama example.

**Comment by geohot** [comment]
I would merge `test_speed_llama.py` that uses fake weights.

**Comment by geohot** [comment]
Profiler adds a lot of overhead, but with NOPROFILE=1 it's pretty close to real perf. #1046

**Comment by roelofvandijk** [comment]
master (52b7105f8798d88feed0197691391f7b9aa4b013)
```
using <class 'tinygrad.runtime.ops_gpu.CLCodegen'>
testing llama python run time
built model
assigned empty tensors, doing warmup
codegen        runtime (median):  211.05ms, runs:   291.58,  210.03,  207.10,  205.42,  213.85,  255.18,  208.70,  205.04,  212.07,  220.11
methodcache    runtime (median):  216.98ms, runs:   219.24,  268.04,  210.51,  204.78,  244.21,  227.40,  204.73,  243.30,  212.41,  214.72
profile        runtime (median): 1106.56ms, runs:  1291.20, 1067.84, 1067.99, 1124.23, 1129.57, 1105.92, 1086.90, 1107.20, 1192.22, 1091.59
```





this branch
```
using <class 'tinygrad.runtime.ops_gpu.CLCodegen'>
testing llama python run time
built model
assigned empty tensors, doing warmup
codegen        runtime (median):  210.12ms, runs:   266.01,  197.88,  209.85,  210.40,  208.92,  207.64,  242.84,  213.48,  214.12,  203.57
methodcache    runtime (median):  211.80ms, runs:   224.87,  267.43,  214.14,  215.10,  198.85,  196.84,  197.96,  209.46,  239.05,  203.25
profile        runtime (median): 1072.59ms, runs:  1258.07, 1057.97, 1113.23, 1064.69, 1139.15, 1069.58, 1070.75, 1072.57, 1078.23, 1072.60
```

**Comment by roelofvandijk** [comment]
> This good to merge?

Yes

**Pull request merged**